File: packages/protocol/src/ice-servers.ts
/**
 * ICE server configuration — the TypeScript counterpart of `packages/wire/iceservers.go`.
 *
 * The signaling server publishes the operator's ICE servers in `/config.json` as an array of
 * `ICEEntry`; the web app parses and validates them with the same rules as the CLI so both
 * clients select direct-path candidates against the same logical set.
 */

/** One STUN/TURN server. Mirrors the browser's `RTCIceServer` and Go's `wire.ICEEntry`. */
export interface ICEEntry {
  /** One or more STUN/TURN URLs that refer to the same server/credentials. */
  urls: string[];
  /** Username for TURN; absent for plain STUN. */
  username?: string;
  /** Credential for TURN; absent for plain STUN. */
  credential?: string;
  /** "password" (default) or "oauth". */
  credentialType?: string;
}

const ICE_SCHEMES = new Set(['stun', 'stuns', 'turn', 'turns']);

/**
 * Validate and fold a single ICE URL into an {@link ICEEntry}. Throws on malformed or unsafe
 * URLs (unknown scheme, empty host, missing/invalid port). STUN entries carry no credentials.
 */
export function parseIceServer(raw: string): ICEEntry {
  const trimmed = raw.trim();
  const schemeEnd = trimmed.indexOf(':');
  if (schemeEnd <= 0) throw new Error(`ice: unsupported scheme in "${raw}"`);
  const scheme = trimmed.slice(0, schemeEnd).toLowerCase();
  if (!ICE_SCHEMES.has(scheme)) {
    throw new Error(`ice: unsupported scheme "${scheme}" in "${raw}" (want stun/stuns/turn/turns)`);
  }

  // Strip credentials in the opaque form (user:cred@host:port) or authority userinfo (//user:cred@host:port).
  let rest = trimmed.slice(schemeEnd + 1);
  if (rest.startsWith('//')) rest = rest.slice(2);
  const at = rest.lastIndexOf('@');
  const userInfo = at >= 0 ? rest.slice(0, at) : '';
  if (at >= 0) rest = rest.slice(at + 1);

  const hostPort = splitHostPort(rest);
  if (hostPort.host === '' || hostPort.port === '') {
    throw new Error(`ice: missing or invalid host:port in "${raw}"`);
  }

  if (scheme === 'turn' || scheme === 'turns') {
    let username = '';
    let credential = '';
    if (userInfo !== '') {
      const i = userInfo.indexOf(':');
      if (i >= 0) {
        username = userInfo.slice(0, i);
        credential = userInfo.slice(i + 1);
      } else {
        username = userInfo;
      }
    }
    const entry: ICEEntry = { urls: [raw] };
    if (username !== '') {
      entry.username = username;
      entry.credential = credential;
    }
    return entry;
  }
  return { urls: [raw] };
}

/** Fold a list of ICE URLs into entries, grouping credential-less STUN URLs together. */
export function parseIceServers(urls: string[]): ICEEntry[] {
  const entries: ICEEntry[] = [];
  for (const raw of urls) {
    if (raw.trim() === '') continue;
    const entry = parseIceServer(raw);
    const last = entries[entries.length - 1];
    if (last !== undefined && sameCreds(last, entry)) {
      last.urls.push(...entry.urls);
    } else {
      entries.push(entry);
    }
  }
  return entries;
}

function sameCreds(a: ICEEntry, b: ICEEntry): boolean {
  const aScheme = a.urls[0]?.split(':')[0]?.toLowerCase();
  const bScheme = b.urls[0]?.split(':')[0]?.toLowerCase();
  return aScheme === bScheme && a.username === b.username && a.credential === b.credential;
}

/**
 * Convert a validated {@link ICEEntry} list into browser `RTCIceServer` objects for
 * `RTCPeerConnection`. Returns undefined when there is nothing to override (caller keeps its
 * bundled defaults).
 */
export function toRTCIceServers(entries: ICEEntry[]): RTCIceServer[] | undefined {
  if (entries.length === 0) return undefined;
  return entries.map((e) => {
    const s: RTCIceServer = { urls: e.urls };
    if (e.username !== undefined && e.username !== '') {
      s.username = e.username;
      s.credential = e.credential ?? '';
    }
    return s;
  });
}

function splitHostPort(rest: string): { host: string; port: string } {
  // IPv6 literals are bracketed: [::1]:3478. Everything else splits on the last ':'.
  if (rest.startsWith('[')) {
    const end = rest.indexOf(']');
    if (end === -1) return { host: '', port: '' };
    const host = rest.slice(1, end);
    const port = rest[end + 1] === ':' ? rest.slice(end + 2) : '';
    return { host, port };
  }
  const i = rest.lastIndexOf(':');
  if (i === -1) return { host: rest, port: '' };
  return { host: rest.slice(0, i), port: rest.slice(i + 1) };
}

File: packages/wire/iceservers.go
package wire

import (
	"fmt"
	"net"
	"net/url"
	"strconv"
	"strings"
)

// ICEEntry describes one STUN/TURN server used to gather ICE candidates for the direct
// path. It is the Go counterpart of the browser's RTCIceServer and the JSON shape the
// signaling server publishes in /config.json, so the CLI and the browser select candidates
// against the same logical set.
type ICEEntry struct {
	// URLs lists one or more STUN/TURN URLs that refer to the same server/credentials.
	URLs []string `json:"urls"`
	// Username is used with TURN credential types; empty for plain STUN.
	Username string `json:"username,omitempty"`
	// Credential is the shared secret or OAuth access token for TURN; never empty for TURN.
	Credential string `json:"credential,omitempty"`
	// CredentialType is "password" (default) or "oauth".
	CredentialType string `json:"credentialType,omitempty"`
}

// ParseICEServers validates one ICE server URL list and folds them into ICEEntry values,
// grouping consecutive URL strings by the presence of credentials. It rejects malformed or
// unsafe URLs (unknown scheme, empty host, missing or non-numeric port) and keeps
// credential-bearing TURN endpoints distinct from credential-less STUN servers so different
// credentials are never merged. STUN URLs must not carry credentials.
func ParseICEServers(urls []string) ([]ICEEntry, error) {
	entries := make([]ICEEntry, 0, len(urls))
	for _, raw := range urls {
		raw = strings.TrimSpace(raw)
		if raw == "" {
			continue
		}
		entry, err := ParseICEServer(raw)
		if err != nil {
			return nil, err
		}
		if n := len(entries); n > 0 && sameCreds(entries[n-1], entry) {
			entries[n-1].URLs = append(entries[n-1].URLs, entry.URLs...)
		} else {
			entries = append(entries, entry)
		}
	}
	return entries, nil
}

// ParseICEServer parses and validates a single ICE URL into an ICEEntry. ICE URLs may use
// either the authority form (stun://host:port) or the common opaque form (stun:host:port);
// both are accepted and validated.
func ParseICEServer(raw string) (ICEEntry, error) {
	u, err := url.Parse(raw)
	if err != nil {
		return ICEEntry{}, fmt.Errorf("ice: malformed URL %q: %w", raw, err)
	}
	scheme := strings.ToLower(u.Scheme)
	switch scheme {
	case "stun", "stuns":
		if err := validateHostPort(raw); err != nil {
			return ICEEntry{}, fmt.Errorf("ice: %q: %w", raw, err)
		}
		return ICEEntry{URLs: []string{raw}}, nil
	case "turn", "turns":
		host, port, _, err := splitICE(raw)
		if err != nil {
			return ICEEntry{}, fmt.Errorf("ice: %q: %w", raw, err)
		}
		_ = host
		_ = port
		username, credential, uerr := iceUserInfo(raw)
		if uerr != nil {
			return ICEEntry{}, fmt.Errorf("ice: %q: %w", raw, uerr)
		}
		return ICEEntry{URLs: []string{raw}, Username: username, Credential: credential}, nil
	default:
		return ICEEntry{}, fmt.Errorf("ice: unsupported scheme %q in %q (want stun/stuns/turn/turns)", u.Scheme, raw)
	}
}

// validateHostPort checks that an ICE URL carries a host and a numeric port. It accepts both
// the opaque and authority forms.
func validateHostPort(raw string) error {
	_, _, _, err := splitICE(raw)
	return err
}

// splitICE returns the host, port, scheme, and error for an ICE URL in either form.
func splitICE(raw string) (host, port, scheme string, err error) {
	u, perr := url.Parse(raw)
	if perr != nil {
		return "", "", "", perr
	}
	scheme = strings.ToLower(u.Scheme)
	rest := u.Opaque
	if rest == "" {
		rest = u.Host
	}
	// The opaque form can carry user:cred@host:port; the userinfo is not part of host:port.
	if at := strings.LastIndex(rest, "@"); at >= 0 {
		rest = rest[at+1:]
	}
	host, port, err = net.SplitHostPort(rest)
	if err != nil {
		return "", "", "", fmt.Errorf("missing or invalid host:port: %w", err)
	}
	if host == "" {
		return "", "", "", fmt.Errorf("empty host")
	}
	if port == "" {
		return "", "", "", fmt.Errorf("missing port")
	}
	if p, cerr := strconv.Atoi(port); cerr != nil || p < 1 || p > 65535 {
		return "", "", "", fmt.Errorf("invalid port %q", port)
	}
	return host, port, scheme, nil
}

// iceUserInfo extracts the username:credential pair from a TURN URL's opaque or authority
// user info, if present.
func iceUserInfo(raw string) (username, credential string, err error) {
	u, perr := url.Parse(raw)
	if perr != nil {
		return "", "", perr
	}
	if u.User != nil && u.User.Username() != "" {
		username = u.User.Username()
		credential, _ = u.User.Password()
		return username, credential, nil
	}
	// Opaque form: user:cred@host:port appears inside the opaque part.
	rest := u.Opaque
	if rest == "" {
		return "", "", nil
	}
	if at := strings.LastIndex(rest, "@"); at >= 0 {
		credPart := rest[:at]
		if i := strings.Index(credPart, ":"); i >= 0 {
			username = credPart[:i]
			credential = credPart[i+1:]
		} else {
			username = credPart
		}
	}
	return username, credential, nil
}

func sameCreds(a, b ICEEntry) bool {
	sameScheme := len(a.URLs) > 0 && len(b.URLs) > 0 && schemeOf(a.URLs[0]) == schemeOf(b.URLs[0])
	return sameScheme && a.Username == b.Username && a.Credential == b.Credential
}

func schemeOf(raw string) string {
	if i := strings.Index(raw, ":"); i >= 0 {
		return strings.ToLower(raw[:i])
	}
	return ""
}
